tools/document-pii/src/pii_finder.rs

```diff
@@ -0,0 +1,301 @@
+) -> Vec<&'a String> {
+    scoped_paths
+        .iter()
+        .filter(|use_path| {
+            let last_use_path = use_path.split("::").last().unwrap().trim();
+            field_types
+                .iter()
+                .any(|field_type| field_type.trim() == last_use_path)
+        })
+        .collect()
+}
+
+/// This function extracts the type names from a complex type and stores them in a BTreeSet.
+/// It's designed to handle nested generic types, such as `Foo<Bar<Baz>>`, and return ["Foo", "Bar", "Baz"].
+fn get_field_types(path: &Path, segments: &mut BTreeSet<String>) {
+    // Iterating over path segments allows us to handle complex, possibly nested types
+    let mut path_iter = path.segments.iter();
+    if let Some(first_segment) = path_iter.next() {
+        let mut ident = first_segment.ident.to_string();
+
+        // Recursion on AngleBracketed args is necessary for nested generic types
+        if let syn::PathArguments::AngleBracketed(angle_bracketed) = &first_segment.arguments {
+            for generic_arg in angle_bracketed.args.iter() {
+                if let syn::GenericArgument::Type(Type::Path(path)) = generic_arg {
+                    get_field_types(&path.path, segments);
+                }
+            }
+        }
+
+        // Namespace resolution: if a second segment exists, it's part of the first type's namespace
+        if let Some(second_segment) = path_iter.next() {
+            ident.push_str("::");
+            ident.push_str(&second_segment.ident.to_string());
+        }
+        segments.insert(ident);
+    }
+}
+
+/// Checks if an attribute is equal to a specified name and value.
+fn has_attr_value(attr: &Attribute, ident: &str, name: &str, value: &str) -> bool {
+    let meta_list = match &attr.meta {
+        Meta::List(meta_list) => meta_list,
+        _ => return false,
+    };
+
+    // Checks name of attribute, E.g. 'metastructure'
+    if !meta_list.path.is_ident(ident) {
+        return false;
+    }
+
+    // Looks for the pattern '{name} = "{value}"'
+
+    let mut iter = meta_list.tokens.clone().into_iter();
+
+    if !iter.any(|token| matches!(token, TokenTree::Ident(ident) if ident == name)) {
+        return false;
+    };
+
+    if !iter.any(|token| matches!(token, TokenTree::Punct(punct) if punct.as_char() == '=')) {
+        return false;
+    };
+
+    iter.any(
+        |token| matches!(token, TokenTree::Literal(lit_val) if lit_val.to_string() == format!("\"{value}\"")),
+    )
+}
+
+/// Checks if a field has the metastructure "pii" and if it does, if it is equal to any
+/// of the values that the user defines.
+fn has_pii_value(pii_values: &[String], field: &Field) -> bool {
+    field.attrs.iter().any(|attribute| {
+        pii_values
+            .iter()
+            .any(|pii_value| has_attr_value(attribute, "metastructure", "pii", pii_value))
+    })
+}
+
+/// Finds all the pii fields recursively of a given type.
+pub fn find_pii_fields_of_type(
```

Comment:
it fees like this function could be on `TypesAndScopedPaths` struct from `item_collector` module.